### ISSUE TITLE
Scheduler dashboard status showing one big square instead of zoomed out view

### ISSUE THREAD

**Issue by dickreuter** [dask/distributed#3130]
When I run distributed jobs, some of them quite long running (30sec), opening the scheduler dashboard shows nice little squares showing the jobs running in parallel. 

However, after a few seconds it zooms in onto a single square and the whole screen is basically one bug blue square. I then need to refresh the page to zoom out again. 

Why does it do that? How can I avoid it?

**Comment by TomAugspurger** [comment]
Old tasks periodically fall off the status dashboard. I believe the source is in https://github.com/dask/distributed/blob/master/distributed/dashboard/components.py if you think there's a bug (do you?)

The task stream plot has a full history.

**Comment by mrocklin** [comment]
We currently refresh the task stream history if there has been a sufficient
lag.  The timeout for this is hard coded and is something like 5-20 seconds
(I don't recall currently).  Probably it would be nice to learn this rate
over time, using something like an exponentially weighted moving average
(or maybe something that is more robust to a wide difference in task
duration) and use that instead.

**Comment by dickreuter** [comment]
@mrocklin , I agree, this would probably solve the problem. I'll play around with it and see if I can raise a PR.

**Comment by mrocklin** [comment]
Thanks Nicolas

**Comment by mrocklin** [comment]
I think that if you search for clear_interval in either
dashboard/components.py or dashboard/scheduler.py (there is some
unfortunate code duplication in the dashboard code at the moment (see
https://github.com/dask/distributed/issues/3048)) then you will see the
fixed value

**Comment by mrocklin** [comment]
Also, to be clear on the wide variance, this problem is made interesting
because often we have some *very* short tasks (like 1ms) which we wouldn't
want to skew our results.  I wonder if maybe we only consider tasks that
are longer than 5s or something like that.

**Comment by dickreuter** [comment]
Could it make sense to keep the display based on a number of tasks rather than based on time?

**Comment by mrocklin** [comment]
In the interactive case people often do a bit of work, then wait for a long
time as they stare at a plot, and then do another burst of work.  If you
keep around the old rectangles then the new work is very thin on the right
side of the plot, and so not very informative.  I think that we *do* want
to clear out the plot after a sufficiently long delay

**Comment by mrocklin** [comment]
Here is another thought, we could reset the current tasks if the gap
between the start of the current task and the end of the last task is
larger than some multiple of the time span of the currently visualized
tasks.

**Comment by dickreuter** [comment]
It looks like the problem is caused by line 1710 in components.scheduler.py: https://github.com/dask/distributed/blob/master/distributed/dashboard/components/scheduler.py#L1710

How is this function called? There appears to be a dictionary called `applications` where it's part of. Is there any way to pass a parameter to this function?

**Comment by mrocklin** [comment]
My recommendation would be to not pass in a constant value, but instead to change things based on recent activity rather than a fixed time duration.

**Comment by mrocklin** [comment]
So for example maybe we only reset things if the time difference between the end of last entry and this current entry is greater than 5s *and* twice the time span of all of the current rectangles on the plot.

**Comment by dickreuter** [comment]
Makes sense, how can I access this information? Is it available in the status_doc function? In the scheduler object or the extra object? Unfortunatley the code has almost no docstrings.

**Comment by jrbourbeau** [comment]
It looks like the current logic is at 

https://github.com/dask/distributed/blob/7cb76f57701db615ecea18976388c9110009e3a1/distributed/dashboard/components/scheduler.py#L902-L909

My guess is that's where we would update the logic for clearing the task stream to use recent activity instead of the hard-coded `self.clear_interval`

**Comment by dickreuter** [comment]
Is there any docstrings at all what first_end, rectangles, self.last etc means? That would be so incredibly useful...

**Comment by mrocklin** [comment]
> Is there any docstrings at all what first_end, rectangles, self.last, self.offset, etc means? 

I suspect that you've already looked and haven't found any.

> That would be so incredibly useful...

I agree.  If you have time to add these that would be welcome.

> How can I get the gap of the current to the last?

I don't recall.  You might have to look at some code here to find out.  I think that you probably want to look at `distributed/diagnostics/task_stream.py`, but it's been a while.

**Comment by dickreuter** [comment]
Ok, I may have found a solution. Updated the PR.
